### PR TITLE
docker_container: fix handling of command and entrypoint in a backwards-compatible way

### LINKED ISSUE #168: container doesn't get recreated when 'command' gets omitted
##### SUMMARY
I need to deploy the same container twice. The first time I use a custom `command` parameter, but the second time want to deploy the container with the default command, so I try to omit the `command` parameter on the second task. Unfortunately, the container doesn't get changed, because this module never detects the change.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

community.docker.docker_container

##### ANSIBLE VERSION

```paste below
ansible 2.9.16
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/christianc/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/christianc/.local/lib/python3.8/site-packages/ansible
  executable location = /data/ssd/christianc/emsconda/envs/core-platform-ng/bin/ansible
  python version = 3.8.10 | emsys distribution | (default, Jun 24 2021, 14:24:42) [GCC 9.3.1 20200408 (Red Hat 9.3.1-2)]

```

##### COLLECTION VERSION

```paste below
ansible-galaxy collection list community.docker
usage: ansible-galaxy collection [-h] COLLECTION_ACTION ...
ansible-galaxy collection: error: argument COLLECTION_ACTION: invalid choice: 'list' (choose from 'init', 'build', 'publish', 'install')
```
Ansible 2.9 doesn't know the command `ansible-galaxy collection list`. But I am using `community.docker` in version 1.7.0.

##### CONFIGURATION

```paste below
DEFAULT_VAULT_IDENTITY_LIST(env: ANSIBLE_VAULT_IDENTITY_LIST) = ['sorry, confidential']
```

##### STEPS TO REPRODUCE

```yaml
- hosts: galera
  tasks:
    - name: Create mariadb bootstrap container
      community.docker.docker_container:
        name: 'mariadb'
        image: 'mariadb:10.5.11'
        restart_policy: always
        network_mode: host
        env:
          MARIADB_ROOT_PASSWORD: '{{ galera_root_password }}'
        log_driver: journald
        command: "{{ '--wsrep-new-cluster' if galera_bootstrap else omit }}"
        container_default_behavior: compatibility
        comparisons:
          'command': strict
      vars:
        galera_bootstrap: true

    - name: Change to non-bootstrap
      community.docker.docker_container:
        name: 'mariadb'
        image: 'mariadb:10.5.11'
        restart_policy: always
        network_mode: host
        env:
          MARIADB_ROOT_PASSWORD: '{{ galera_root_password }}'
        log_driver: journald
        command: "{{ '--wsrep-new-cluster' if galera_bootstrap else omit }}"
        container_default_behavior: compatibility
        comparisons:
          'command': strict
      vars:
        galera_bootstrap: false

```
The example is a bit more complicated than it needs to be. You can just as well use `command: "--wsrep-new-cluster"` on the first task and omit this parameter on the second task completely. 

Not matter what I try, this module never changes the container when executing the second task.

##### EXPECTED RESULTS
Recreated container with custom `command` removed.

##### ACTUAL RESULTS
Second tasks return with `"changed": false`.

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Fixes #168.

Since this would be a breaking change, I'm adding a new option whose current default value is deprecated and will change in 2.0.0. The deprecation warning is only emitted if there would actually be a difference in behavior.

##### ISSUE TYPE
- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME
docker_container

### CODE REVIEW

**Comment by ChristianCiach** [comment]
> When this is set to C(correct), these options are converted to strings considering shell quoting rules, and an empty value or empty list will be handled for idempotency checks, with the meaning "use default value from image".

Maybe I do not understand this correctly, but does that mean that it will be impossible to set the entrypoint or command to an *actual* empty value? Overriding the default entrypoint of an image with an empty value is a real use case.

**Comment by felixfontein** [comment]
> > When this is set to C(correct), these options are converted to strings considering shell quoting rules, and an empty value or empty list will be handled for idempotency checks, with the meaning "use default value from image".
> 
> Maybe I do not understand this correctly, but does that mean that it will be impossible to set the entrypoint or command to an _actual_ empty value? Overriding the default entrypoint of an image with an empty value is a real use case.

I think my statement is wrong. Empty values should do what you expect. From how I understand the Docker daemon code, the behavior is as follows:
1. Container config values for command and entrypoint can be `nil` or a string array (https://github.com/docker/engine/blob/f07e53e0bb97c6364032bb14a020c50118eb7394/api/types/container/config.go#L55-L61, https://github.com/moby/moby/blob/f07e53e0bb97c6364032bb14a020c50118eb7394/api/types/strslice/strslice.go);
2. On merge (https://github.com/docker/engine/blob/f07e53e0bb97c6364032bb14a020c50118eb7394/daemon/commit.go#L68-L76):
   - If entrypoint is non-empty, both command and entrypoint are used.
   - If both entrypoint and command are `nil` or empty, the image's command is used.
   - If entrypoint is `nil`, the image's entrypoint is used.
3. After merging, if the entrypoint has length 1 and contains only an empty string, it is set to `nil` (https://github.com/docker/engine/blob/f07e53e0bb97c6364032bb14a020c50118eb7394/daemon/create.go#L277-L280).

(If someone is interested, https://github.com/docker/engine/blob/f07e53e0bb97c6364032bb14a020c50118eb7394/daemon/container.go#L190-L195 combines entrypoint and command to what is actually executed.)

So the code in the PR still allows to pass an empty list / string to docker daemon (that was possible before), and as opposed to the old code, it handles empty list / string correctly for idempotency (by not interpreting it as "not specified").

**Comment by ChristianCiach** [comment]
Excellent. This is exactly how I think this should behave. Thanks!

**Comment by felixfontein** [comment]
> 3. After merging, if the entrypoint has length 1 and contains only an empty string, it is set to `nil` (https://github.com/docker/engine/blob/f07e53e0bb97c6364032bb14a020c50118eb7394/daemon/create.go#L277-L280).

This case got me thinking, since right now it is not possible to pass this on to the Docker daemon, since the module converts everything to strings first.

I think I'll change it so that it works with lists instead, and does the list → string → list conversion for `command_handling=compatibility` for backwards compatibility.

**Comment by felixfontein** [comment]
ready_for_review

**Review by briantist — APPROVED** [review]
This looks good to me! The change in behavior makes sense, the documentation is clear. There's a way forward to pre-emptively choose the "correct" behavior in anticipation of a new version and advance warning of the default change. Tests look great and seem to cover all the bases.

As discussed in IRC, Ill leave it up to you to decide when to make the breaking change to the default value.

File: tests/integration/targets/docker_container/tasks/tests/options.yml
Comment (felixfontein) [reply tests/integration/targets/docker_container/tasks/tests/options.yml:1522]: I'm tending to keep the tests as-is. If someone wants to rework all the docker_container tests, I won't stop them. Warning: it's a **lot** of work :)

**Comment by felixfontein** [comment]
@ChristianCiach @briantist thanks for your comments and reviews!

**Pull request merged**
